### PR TITLE
[CALCITE-2591] EnumerableDefaults#mergeJoin should throw error and not return incorrect results when inputs are not ordered

### PR DESCRIPTION
**Pull request by eolivelli**

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 1]
```diff
@@ -1371,8 +1371,12 @@ private void closeInner() {
       final Function2<TSource, TInner, TResult> resultSelector,
       boolean generateNullsOnLeft,
       boolean generateNullsOnRight) {
-    assert !generateNullsOnLeft : "not implemented";
-    assert !generateNullsOnRight : "not implemented";
+    if (generateNullsOnLeft) {
+      throw new UnsupportedOperationException("not implemented");
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: Please use meaningful exception messages.

Do you really want to see `UnsupportedOperationException("not implemented")` in production logs? I bet you don't

Comment (eolivelli) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: done

**Comment by eolivelli** [comment]
@vlsi fixed the message, please take another look

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 2]
```diff
@@ -3263,7 +3267,10 @@ private boolean advance() {
         TKey leftKey2 = outerKeySelector.apply(left);
         int c = leftKey.compareTo(leftKey2);
         if (c != 0) {
-          assert c < 0 : "not sorted";
+          if (c < 0) {
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: The condition should be `c > 0`.
Original assertion was `c<0==true`, and assertion is violated in `c>0` case.

Comment (eolivelli) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: oh, yes, sorry, trivial error !!!
So it should be "c >= 0" strictly speaking

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 3]
```diff
@@ -3263,7 +3267,10 @@ private boolean advance() {
...
+            throw new IllegalStateException(
+              "not sorted, " + leftKey + " is less than " + leftKey2);
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: The message should be updated to account `c>0`.
A bit more verbose message won't hurt.

For instance: "mergeJoin assumes sorted in ascending order input, however ... "

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 2 above)
Comment (vlsi) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: `c==0` is acceptable.

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java [hunk 4]
```diff
@@ -3263,7 +3269,11 @@ private boolean advance() {
         TKey leftKey2 = outerKeySelector.apply(left);
         int c = leftKey.compareTo(leftKey2);
         if (c != 0) {
-          assert c < 0 : "not sorted";
+          if (c > 0) {
+            throw new IllegalStateException(
+              "mergeJoin assumes sorted in ascending order input, "
+                 + "however " + leftKey + " is less than " + leftKey2);
```
Comment (vlsi) [inline linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: `leftKey + " is less than " + leftKey2` seems to be wrong.

`c = leftKey.compareTo(leftKey2)`.
`c > 0 ` ==> `leftKey > leftKey2`

**Comment by eolivelli** [comment]
Thank you @vlsi for your quick review

I have applied your suggestions

File: linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java (hunk 4 above)
Comment (eolivelli) [reply linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java]: yes, sure, I wrote the comment when the assertion was wrong :-)
sorry

**Comment by vlsi** [comment]
LGTM.

Will commit if no-one objects in a day or two.

**Comment by eolivelli** [comment]
Thank you very much.
If you have time please take a look to my real problem on the mailing list

**Pull request closed without merging**